walkdir 2.0

behinger opened the pull request:
closes #3 - I decided to simply redo the PR due to the formatting issue.

This should be fixed, I had "auto-format" activated in my IDE.

aplavin committed:
add show(), extend joinpath()

aplavin committed:
don't use joinpath()

aplavin committed:
add & upd compat

aplavin committed:
support more funcs

behinger committed:
walkdir branch, with review-fixes but no formatting

behinger commented:
argh -.- Formatting crept back in - whyyyy.. sorry!

behinger committed:
undo formatting

behinger committed:
undo more formatting

behinger commented:
ok. This should do it - sorry for the messy PR.

aplavin commented:
Nice, thanks! Happy to accept, just add some tests – both for public projects without token and for walkdir.

behinger committed:
public download added, unit tests

behinger commented:
- I added a few unittests
- I added download functionality. now URL first checks if there is a public download link available, and only then goes to the view link. This was a surprisingly surgical addon
- I added a cp functionality for dirs, mirroring what julia does (always recursive, using mkpath and looping recursively through `readdir`)

todo:
- [ ] check compatibility with private project (I dont have an API key available right now)

aplavin commented on src/highlevel.jl:
```diff
@@ -226,7 +241,57 @@ versions(f::File) = [
-url(f) = url(f, only(view_only_links(project(f))))
+function url(f)
+    try 
+        return f.entity.links[:download]  # public open link
```
I tried for a private project, and `f.entity.links[:download]` always exists. It just points to an HTML page.
Is there another way to determine the presence of a public download link?

behinger replied:
Ah that is very unfortunate. I guess the :download for a private project would work too, if the appropriate header is send? I unfortunately don't know much about API usage, but that should work right?

Or is the problem that the API key is not available in the single files and therefore the view link should be used?

aplavin replied:
I looked into it a long time ago originally, and the only way I found to download files from private projects was through these "view-only links". Maybe there's a cleaner way indeed, looking forward to code suggestions!

For now, we can use `public` attribute to distinguish, I guess... Will try doing that.

aplavin commented:
Thanks, included these changes with some modifications. Please try and check that all necessary functionality is present now, and I'll release the new version.

behinger commented:
looks good to me! I'll try to use it in a package soon and will report :-)
